### ISSUE TITLE
SyntaxNode.{Replace,Remove}Node{,s} loses syntax tree options

### ISSUE THREAD

**Issue by jskeet** [dotnet/roslyn#19741]
(As originally reported in https://stackoverflow.com/questions/44060524; @jasonmalinowski suggested it was worth reporting here.)

**Version Used**: 2.1.0

**Steps to Reproduce**:

Code, demonstrating replacing a node (but other calls have the same issue):

```csharp
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Scripting;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Scripting;
using System;
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        Script script = CSharpScript.Create("Console.WriteLine(\"Before\")",
            ScriptOptions.Default.AddImports("System"));

        var compilation = script.GetCompilation();
        var tree = compilation.SyntaxTrees.Single();

        var after = SyntaxFactory.LiteralExpression(
            SyntaxKind.StringLiteralExpression,
            SyntaxFactory.Literal("After"));

        var root = tree.GetRoot();
        var before = root.DescendantNodes().OfType<LiteralExpressionSyntax>().Single();
        var newRoot = root.ReplaceNode(before, after);
        var fixedTree = newRoot.SyntaxTree.WithRootAndOptions(newRoot, tree.Options);

        Console.WriteLine(newRoot);                         // Console.WriteLine("After")
        Console.WriteLine(tree.Options.Kind);               // Script
        Console.WriteLine(newRoot.SyntaxTree.Options.Kind); // Regular
        Console.WriteLine(fixedTree.Options.Kind);          // Script
    }
}
```

Project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net462</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.1.0" />
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="2.1.0" />
  </ItemGroup>

</Project>
```

**Expected Behavior**:

`ReplaceNode` maintains the options in the syntax tree

**Actual Behavior**:

Options are reset to default.

@filipw has worked out that this is the responsible code:

https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs#L344-L357

Questions:

- Is this by design for some reason we can't fathom?
- Would fixing it now be too much of a breaking change? (Would anyone really be relying on this?)

**Comment by jcouv** [comment]
FYI @gafter

**Comment by jcouv** [comment]
From discussion with @VSadov this behavior is by default. When calling `ReplaceNode` you get a new "floating" node (ie. not attached to any `SyntaxTree`).
If you ask for its `SyntaxTree`, we could have returned `null`, but for some reason that I do not know, we chose to make a dummy tree on the spot. That seems convenient in some common cases, but seems confusing and error-prone too... We probably cannot return `null` at this point for compat reasons.

The proper pattern would be to call `ReplaceNode` and then explicitly make your own `SyntaxTree` to hold it.

Ping @gafter in case anything to add.

**Comment by gafter** [comment]
The design of the trees is such that every node is part of some tree. We create a tree lazily if we do not have any hint where to get options from.

Having said that, this is a bug. We have tree options that we should be preserving into the new tree in this case.

**Comment by jcouv** [comment]
Fix merged for 15.6. Thanks
